Advanced scanOnly scenario to scan existing loadmodules

dennis-behm opened the pull request:
This is the PR for issue: #85 

With this enhancement, it is possible to populate the _collection for build outputs (aka output collection)_ in a scanOnly scenario without rebuilding the modules. This is crucial for migration scenarios, which include **static linkage** and would like to skip a read full build at the beginning. 

It introduces a new cmdline option `--scanLoadmodules` which can only be used along with `-scanOnly`.

Sample invocation to invoke a scanOnly and scanLoadmodules:
```
groovyz /u/dbehm/test/dbb-zappbuild/build.groovy --sourceDir /u/dbehm/test/dbb-zappbuild/samples --workDir /u/dbehm/test/work/Mortgage --hlq DBEHM.DBB.BUILD --application MortgageApplication --scanOnly --fullBuild --scanLoadmodules
```

Sample output. Note: Output files not being  found will be documented.
```
** Build start at 20210223.094610.046
** Repository client created for https://10.3.20.96:10443/dbb
** Build output located at /u/dbehm/test/work/Mortgage/build.20210223.094610.046
** Build result created for BuildGroup:MortgageApplication-scanLoadModule BuildLabel:build.20210223.094610.046 at https://10.3.20.96:10443/dbb/rest/buildResult/39421
** --fullBuild option selected. Scanning all programs for application MortgageApplication
** Writing build list file to /u/dbehm/test/work/Mortgage/build.20210223.094610.046/buildList.txt
** Scanning loadmodules for static dependencies.
!* Error scanning output file for MortgageApplication/cobol/epscsmrd.cbl  : DBEHM.DBB.BUILD.LOAD(EPSCSMRD)
com.ibm.dbb.build.ValidationException: BGZTK0046E Could not scan load module //'DBEHM.DBB.BUILD.LOAD(EPSCSMRD)' because it does not exist.
** Writing build report data to /u/dbehm/test/work/Mortgage/build.20210223.094610.046/BuildReport.json
** Writing build report to /u/dbehm/test/work/Mortgage/build.20210223.094610.046/BuildReport.html
** Build ended at Tue Feb 23 09:46:28 GMT+01:00 2021
** Build State : CLEAN
** Total files processed : 15
** Total build time  : 17.894 seconds
```

jbyibm commented on utilities/ImpactUtilities.groovy:
```diff
@@ -199,6 +199,66 @@ def calculateChangedFiles(BuildResult lastBuildResult) {
+	}
+}
+
+/*
+ * returns languagePrefix for language script name or null if not defined.
+ */
+def getLangPrefix(String scriptName){
+	def langPrefix = null
+	switch(scriptName) {
+		case "Cobol.groovy":
+			langPrefix = 'cobol'
+			break;
+		case "LinkEdit.groovy" :
+			langPrefix = 'linkedit'
+			break;
+		case "PLI.groovy":
+			langPrefix = 'pli'
+			break;
+		case "Assembler.groovy":
+			langPrefix = 'assembler'
+			break;
+		default:
+			if (props.verbose) println ("*** No language prefix defined for $scriptName.")
+			break;
+	}
+	return langPrefix
+}
```
This method may go in BuildUtilities.groovy?

jbyibm reviewed (commented):
Just some minor comments

drbruce-git commented on BUILD.md:
```diff
@@ -61,6 +65,7 @@ build options:
  -i,--impactBuild         Flag indicating to build only programs impacted
                           by changed files since last successful build.
  -s,--scanOnly            Flag indicating to only scan files for application
+ -sl,--scanLoadmodules    Flag indicating to scan loadmodules based on provided hlq, mutual to --scanOnly
```
How about introducing some new scan options:
```
-s,--scanOnly             Flag indicating to only scan source files for application without building anything (deprecated use --scanSource)
-ss,--scanSource          Flag indicating to only scan source files for application without building anything
-sl,--scanLoad            Flag indicating to only scan load modules for application without building anything
-sa,--scanAll             Flag indicating to scan both source files and load modules for application without building anything
```

dennis-behm commented:
@jbyibm @drbruce-git

I have included your comments and have further refined the new logic.

- Added the new command line options.
- Moved to the scanner API, rather than using the DependencyResolver. Comments?
- Fixed a defect about link cards, which don't carry the `_linkEdit` file property, but need to be included in this logic.

jbyibm reviewed (approved):
Look good

dennis-behm commented:
Due to the feedback of the reviewers, the command line option has slightly changed. `build.groovy` now accepts a new set of options:

```
  -s,--scanOnly            Flag indicating to only scan source files for application without building anything (deprecated use --scanSource)
  -ss,--scanSource         Flag indicating to only scan source files for application without building anything
  -sl,--scanLoad           Flag indicating to only scan load modules for application without building anything
  -sa,--scanAll            Flag indicating to scan both source files and load modules for application without building anything
 ```

### Sample scenario
- The existing load modules for my application reside on `DBEHM.MIGRATE.BUILD`
- I already have moved my source code to git. My application uses has static links.
- I would like to avoid the `fullBuild` to setup the output collection.

To setup the source and output collection, I am using the new option `--scanAll` - line breaks are added only for better readability: 
```
/usr/lpp/dbb/v1r0/bin/groovyz /u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/build.groovy
    --sourceDir /u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples
    --workDir /u/dbehm/test/work/Mortgage
    --hlq DBEHM.MIGRATE.BUILD
    --application MortgageApplication
    --scanAll --fullBuild
```
The new option will also scan the load modules for static dependencies:
```
** Build start at 20210225.021541.015
** Repository client created for https://10.3.20.96:10443/dbb
** Build output located at /u/dbehm/test/work/Mortgage/build.20210225.021541.015
** Build result created for BuildGroup:MortgageApplication-scanLoadModule BuildLabel:build.20210225.021541.015 at https://10.3.20.96:10443/dbb/rest/buildResult/40154
** --fullBuild option selected. Scanning all programs for application MortgageApplication
** Writing build list file to /u/dbehm/test/work/Mortgage/build.20210225.021541.015/buildList.txt
** Scanning source code.
** Scanning load modules for static dependencies.
** Writing build report data to /u/dbehm/test/work/Mortgage/build.20210225.021541.015/BuildReport.json
** Writing build report to /u/dbehm/test/work/Mortgage/build.20210225.021541.015/BuildReport.html
** Build ended at Thu Feb 25 14:15:51 GMT+01:00 2021
** Build State : CLEAN
** Total files processed : 15
** Total build time  : 9.975 seconds
```
With `--verbose` you will see all the output as:
```
** Repository client created for https://10.3.20.96:10443/dbb
** Build output located at /u/dbehm/test/work/Mortgage/build.20210225.022132.021
** Build result created for BuildGroup:MortgageApplication-scanLoadModule BuildLabel:build.20210225.022132.021 at https://10.3.20.96:10443/dbb/rest/buildResult/40184
** Created collection MortgageApplication-scanLoadModule
** Created collection MortgageApplication-scanLoadModule-outputs
** --fullBuild option selected. Scanning all programs for application MortgageApplication
** Writing build list file to /u/dbehm/test/work/Mortgage/build.20210225.022132.021/buildList.txt
MortgageApplication/copybook/epsmtout.cpy
MortgageApplication/cobol/epsnbrvl.cbl
MortgageApplication/cobol/epscsmrt.cbl
MortgageApplication/bms/epsmort.bms
MortgageApplication/link/epsmlist.lnk
MortgageApplication/copybook/epsmortf.cpy
MortgageApplication/copybook/epsnbrpm.cpy
MortgageApplication/bms/epsmlis.bms
MortgageApplication/copybook/epsmtcom.cpy
MortgageApplication/cobol/epsmlist.cbl
MortgageApplication/copybook/epsmtinp.cpy
MortgageApplication/copybook/epspdata.cpy
MortgageApplication/cobol/epsmpmt.cbl
MortgageApplication/cobol/epscmort.cbl
MortgageApplication/cobol/epscsmrd.cbl
** Scanning source code.
** Updating collections MortgageApplication-scanLoadModule and MortgageApplication-scanLoadModule-outputs
*** Scanning file MortgageApplication/copybook/epsmtout.cpy (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/copybook/epsmtout.cpy)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/copybook/epsmtout.cpy =
{"dli":false,"lname":"EPSMTOUT","file":"MortgageApplication\/copybook\/epsmtout.cpy","cics":false,"language":"COB","sql":false}
*** Scanning file MortgageApplication/cobol/epsnbrvl.cbl (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/cobol/epsnbrvl.cbl)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/cobol/epsnbrvl.cbl =
{"dli":false,"lname":"EPSNBRVL","file":"MortgageApplication\/cobol\/epsnbrvl.cbl","cics":false,"logicalDependencies":[{"lname":"EPSNBRPM","library":"SYSLIB","category":"COPY"}],"language":"COB","sql":false}
*** Scanning file MortgageApplication/cobol/epscsmrt.cbl (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/cobol/epscsmrt.cbl)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/cobol/epscsmrt.cbl =
{"dli":false,"lname":"EPSCSMRT","file":"MortgageApplication\/cobol\/epscsmrt.cbl","cics":false,"logicalDependencies":[{"lname":"EPSMTCOM","library":"SYSLIB","category":"COPY"},{"lname":"EPSPDATA","library":"SYSLIB","category":"COPY"}],"language":"COB","sql":false}
*** Scanning file MortgageApplication/bms/epsmort.bms (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/bms/epsmort.bms)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/bms/epsmort.bms =
{"dli":false,"lname":"EPSMORT","file":"MortgageApplication\/bms\/epsmort.bms","cics":false,"language":"ASM","sql":false}
*** Scanning file MortgageApplication/link/epsmlist.lnk (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/link/epsmlist.lnk)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/link/epsmlist.lnk =
{"dli":false,"lname":"EPSMLIST","file":"MortgageApplication\/link\/epsmlist.lnk","cics":false,"language":"ZBND","sql":false}
*** Scanning file MortgageApplication/copybook/epsmortf.cpy (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/copybook/epsmortf.cpy)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/copybook/epsmortf.cpy =
{"dli":false,"lname":"EPSMORTF","file":"MortgageApplication\/copybook\/epsmortf.cpy","cics":false,"language":"COB","sql":false}
*** Scanning file MortgageApplication/copybook/epsnbrpm.cpy (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/copybook/epsnbrpm.cpy)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/copybook/epsnbrpm.cpy =
{"dli":false,"lname":"EPSNBRPM","file":"MortgageApplication\/copybook\/epsnbrpm.cpy","cics":false,"language":"COB","sql":false}
*** Scanning file MortgageApplication/bms/epsmlis.bms (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/bms/epsmlis.bms)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/bms/epsmlis.bms =
{"dli":false,"lname":"EPSMLIS","file":"MortgageApplication\/bms\/epsmlis.bms","cics":false,"language":"ASM","sql":false}
*** Scanning file MortgageApplication/copybook/epsmtcom.cpy (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/copybook/epsmtcom.cpy)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/copybook/epsmtcom.cpy =
{"dli":false,"lname":"EPSMTCOM","file":"MortgageApplication\/copybook\/epsmtcom.cpy","cics":false,"logicalDependencies":[{"lname":"EPSMTINP","library":"SYSLIB","category":"COPY"},{"lname":"EPSMTOUT","library":"SYSLIB","category":"COPY"}],"language":"COB","sql":false}
*** Scanning file MortgageApplication/cobol/epsmlist.cbl (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/cobol/epsmlist.cbl)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/cobol/epsmlist.cbl =
{"dli":false,"lname":"EPSMLIST","file":"MortgageApplication\/cobol\/epsmlist.cbl","cics":true,"logicalDependencies":[{"lname":"DFHAID","library":"SYSLIB","category":"COPY"},{"lname":"EPSMLIS","library":"SYSLIB","category":"COPY"},{"lname":"EPSMORTF","library":"SYSLIB","category":"COPY"},{"lname":"EPSMTCOM","library":"SYSLIB","category":"COPY"},{"lname":"EPSNBRPM","library":"SYSLIB","category":"COPY"}],"language":"COB","sql":false}
*** Scanning file MortgageApplication/copybook/epsmtinp.cpy (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/copybook/epsmtinp.cpy)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/copybook/epsmtinp.cpy =
{"dli":false,"lname":"EPSMTINP","file":"MortgageApplication\/copybook\/epsmtinp.cpy","cics":false,"language":"COB","sql":false}
*** Scanning file MortgageApplication/copybook/epspdata.cpy (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/copybook/epspdata.cpy)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/copybook/epspdata.cpy =
{"dli":false,"lname":"EPSPDATA","file":"MortgageApplication\/copybook\/epspdata.cpy","cics":false,"language":"COB","sql":false}
*** Scanning file MortgageApplication/cobol/epsmpmt.cbl (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/cobol/epsmpmt.cbl)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/cobol/epsmpmt.cbl =
{"dli":false,"lname":"EPSMPMT","file":"MortgageApplication\/cobol\/epsmpmt.cbl","cics":false,"logicalDependencies":[{"lname":"EPSPDATA","library":"SYSLIB","category":"COPY"}],"language":"COB","sql":false}
*** Scanning file MortgageApplication/cobol/epscmort.cbl (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/cobol/epscmort.cbl)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/cobol/epscmort.cbl =
{"dli":false,"lname":"EPSCMORT","file":"MortgageApplication\/cobol\/epscmort.cbl","cics":true,"logicalDependencies":[{"lname":"DFHAID","library":"SYSLIB","category":"COPY"},{"lname":"EPSMORT","library":"SYSLIB","category":"COPY"},{"lname":"EPSMTCOM","library":"SYSLIB","category":"COPY"},{"lname":"EPSNBRPM","library":"SYSLIB","category":"COPY"},{"lname":"SQLCA","library":"SYSLIB","category":"SQL INCLUDE"}],"language":"COB","sql":true}
*** Scanning file MortgageApplication/cobol/epscsmrd.cbl (/u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication/cobol/epscsmrd.cbl)
*** Scanning file with the default scanner
*** Logical file for MortgageApplication/cobol/epscsmrd.cbl =
{"dli":false,"lname":"EPSCSMRD","file":"MortgageApplication\/cobol\/epscsmrd.cbl","cics":true,"language":"COB","sql":false}
** Storing 15 logical files in repository collection 'MortgageApplication-scanLoadModule'
HTTP/1.1 200 OK
** Scanning load modules for static dependencies.
*** Scanning file with the default scanner
*** Skipped scanning module DBEHM.MIGRATE.BUILD.LOAD(EPSNBRVL) of MortgageApplication/cobol/epsnbrvl.cbl.
*** Scanning file with the default scanner
*** Scanning load module DBEHM.MIGRATE.BUILD.LOAD(EPSCSMRT) of MortgageApplication/cobol/epscsmrt.cbl
*** Scanning load module for MortgageApplication/cobol/epscsmrt.cbl
*** Logical file = 
{"dli":false,"lname":"EPSCSMRT","file":"MortgageApplication\/cobol\/epscsmrt.cbl","cics":false,"language":"ZBND","sql":false}
*** No language prefix defined for BMS.groovy.
*** Skipped scanning outputs of MortgageApplication/bms/epsmort.bms. No language prefix found.
*** Scanning file with the default scanner
*** Scanning load module DBEHM.MIGRATE.BUILD.LOAD(EPSMLIST) of MortgageApplication/link/epsmlist.lnk
*** Scanning load module for MortgageApplication/link/epsmlist.lnk
*** Logical file = 
{"dli":false,"lname":"EPSMLIST","file":"MortgageApplication\/link\/epsmlist.lnk","cics":false,"logicalDependencies":[{"lname":"EPSMPMT","library":"DBEHM.MIGRATE.BUILD.LOAD","category":"LINK"},{"lname":"EPSMLIST","library":"DBEHM.MIGRATE.BUILD.OBJ","category":"LINK"}],"language":"ZBND","sql":false}
*** No language prefix defined for BMS.groovy.
*** Skipped scanning outputs of MortgageApplication/bms/epsmlis.bms. No language prefix found.
*** Scanning file with the default scanner
*** Skipped scanning module DBEHM.MIGRATE.BUILD.LOAD(EPSMLIST) of MortgageApplication/cobol/epsmlist.cbl.
*** Scanning file with the default scanner
*** Scanning load module DBEHM.MIGRATE.BUILD.LOAD(EPSMPMT) of MortgageApplication/cobol/epsmpmt.cbl
*** Scanning load module for MortgageApplication/cobol/epsmpmt.cbl
*** Logical file = 
{"dli":false,"lname":"EPSMPMT","file":"MortgageApplication\/cobol\/epsmpmt.cbl","cics":false,"language":"ZBND","sql":false}
*** Scanning file with the default scanner
*** Scanning load module DBEHM.MIGRATE.BUILD.LOAD(EPSCMORT) of MortgageApplication/cobol/epscmort.cbl
*** Scanning load module for MortgageApplication/cobol/epscmort.cbl
*** Logical file = 
{"dli":false,"lname":"EPSCMORT","file":"MortgageApplication\/cobol\/epscmort.cbl","cics":false,"logicalDependencies":[{"lname":"EPSNBRVL","library":"DBEHM.MIGRATE.BUILD.OBJ","category":"LINK"}],"language":"ZBND","sql":false}
*** Scanning file with the default scanner
*** Scanning load module DBEHM.MIGRATE.BUILD.LOAD(EPSCSMRD) of MortgageApplication/cobol/epscsmrd.cbl
*** Scanning load module for MortgageApplication/cobol/epscsmrd.cbl
*** Logical file = 
{"dli":false,"lname":"EPSCSMRD","file":"MortgageApplication\/cobol\/epscsmrd.cbl","cics":false,"language":"ZBND","sql":false}
** Writing build report data to /u/dbehm/test/work/Mortgage/build.20210225.022132.021/BuildReport.json
** Writing build report to /u/dbehm/test/work/Mortgage/build.20210225.022132.021/BuildReport.html
** Updating build result BuildGroup:MortgageApplication-scanLoadModule BuildLabel:build.20210225.022132.021
*** Obtaining hash for directory /u/dbehm/test/dbb-zappbuild-test/dbb-zappbuild/samples/MortgageApplication
** Setting property :githash:MortgageApplication : 1cebdb5523ecfaee90aeab191615e9c678dc003e
** Build ended at Thu Feb 25 14:21:50 GMT+01:00 2021
** Build State : CLEAN
** Total files processed : 15
** Total build time  : 17.443 seconds
```

After changing `epsmlist.cbl`, the impact build will understand that link card `epsmlist.lnk` needs to be processed, because the subpgm has changed.
```
** Build start at 20210225.022155.021
** Repository client created for https://10.3.20.96:10443/dbb
** Build output located at /u/dbehm/test/work/Mortgage/build.20210225.022155.021
** Build result created for BuildGroup:MortgageApplication-scanLoadModule BuildLabel:build.20210225.022155.021 at https://10.3.20.96:10443/dbb/rest/buildResult/40204
** --impactBuild option selected. Building impacted programs for application MortgageApplication 
** Writing build list file to /u/dbehm/test/work/Mortgage/build.20210225.022155.021/buildList.txt
** Invoking build scripts according to build order: BMS.groovy,Cobol.groovy,LinkEdit.groovy
** Building files mapped to Cobol.groovy script
*** Building file MortgageApplication/cobol/epsmlist.cbl
** Building files mapped to LinkEdit.groovy script
*** Building file MortgageApplication/link/epsmlist.lnk
** Writing build report data to /u/dbehm/test/work/Mortgage/build.20210225.022155.021/BuildReport.json
** Writing build report to /u/dbehm/test/work/Mortgage/build.20210225.022155.021/BuildReport.html
** Build ended at Thu Feb 25 14:22:02 GMT+01:00 2021
** Build State : CLEAN
** Total files processed : 2
** Total build time  : 7.485 seconds
```

dennis-behm commented:
With this enhancement, there are also a couple of topics to consider:

Now we can run `impactBuilds` in static linkage scenarios after migrating source code to git, but builds potentially might fail, because objects are not on the libraries .

So it requires some thinking how to use it. Most likely use the SYSLIB concatenation for OBJECT to concatenate with the current production library: 
https://github.com/IBM/dbb-zappbuild/blob/development/samples/application-conf/Cobol.properties#L65-L67
